src/neptune/api/searching_entries.py

```diff
@@ -148,34 +150,66 @@ def iter_over_pages(
     max_offset: int = MAX_SERVER_OFFSET,
     sort_by_column_type: Optional[str] = None,
     ascending: bool = False,
+    progress_bar: Optional[Union[bool, Type[ProgressBarCallback]]] = None,
     **kwargs: Any,
 ) -> Generator[Any, None, None]:
     searching_after = None
     last_page = None
 
-    while True:
-        if last_page:
-            page_attribute = find_attribute(entry=last_page[-1], path=sort_by)
+    total = get_single_page(
+        limit=0,
+        offset=0,
+        **kwargs,
+    ).get("matchingItemCount", 0)
 
-            if not page_attribute:
-                raise ValueError(f"Cannot find attribute {sort_by} in last page")
+    progress_bar = progress_bar if step_size >= total else None
+    extracted_records = 0
 
-            searching_after = page_attribute.properties["value"]
+    with _construct_progress_bar(progress_bar) as bar:
+        while True:
+            if last_page:
+                page_attribute = find_attribute(entry=last_page[-1], path=sort_by)
+
+                if not page_attribute:
+                    raise ValueError(f"Cannot find attribute {sort_by} in last page")
+
+                searching_after = page_attribute.properties["value"]
+
+            for offset in range(0, max_offset, step_size):
+                result = get_single_page(
+                    limit=min(step_size, max_offset - offset),
+                    offset=offset,
+                    sort_by=sort_by,
+                    sort_by_column_type=sort_by_column_type,
+                    searching_after=searching_after,
+                    ascending=ascending,
+                    **kwargs,
+                )
+
+                page = _entries_from_page(result)
+
+                if not page:
+                    return
+
+                extracted_records += len(page)
+                if extracted_records >= max_offset:  # new 10k page started
```

Comment:
I don't understand. To my knowledge, this will never be reached as page size is at most `min(step_size, max_offset -offset)` so it will never reach `max_offset`.